modules/packages/Channel.chpl

```diff
@@ -0,0 +1,574 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+This module contains the implementation of channels which can be used to move
+typed data between Chapel tasks.
+
+A channel is a parallel-safe data structure that provides a mechanism for
+concurrently executing functions to communicate by sending and receiving values
+of a specified element type. A channel can be buffered or unbuffered. A
+buffered channel has a maximum capacity specified by ``bufferSize``. There are
+mainly three operations that can be performed on the channel.
+
+    * :proc:`send` : Send a value to the channel.
+    * :proc:`recv` : Receive a value from the channel.
+    * :proc:`close` : Close a channel such that no more value can be sent to it.
+
+The channel operations are blocking, i.e., the calling task will be suspended
+if an operation cannot be completed. The channel follows First-In-First-Out
+mechanism, i.e., the first value sent to the channel will be received first.
+*/
+
+module Channel {
+
+    use CPtr;
+    use Sort;
+    use SysCTypes;
+    use List;
+    use Random;
+
+    pragma "no doc"
+    class Waiter {
+        // Class used to maintain the properties of suspended tasks.
+
+        type valueType;
+        var val : c_ptr(valueType);
```

Comment:
Is the use of `c_ptr` here a workaround for not having `ref` fields? If so, can you add a comment saying that?